### ISSUE TITLE
API changes

### ISSUE THREAD

**Issue by rarbg** [rarbg/torrentapi#1]
@abeloin , @fernandog , @SchizoDuckie , @Taloth , @JackDandy , @cvium , @zone117x , @mancuniancol , @romanvm , @Hellowlol , @erickok , @KrunchMuffin 
Guys.. i took the liberty of creating this project , i dont know if you know each other so i created the README with all of you listed. Now lets make the api the way you like it.
Everyone has different issues related to the api so just please tell me how you want it to respond to your requests , what extra features you like or what you want changed and ill make it happen asap

Changes so far :

- [ ] Make per app api hit statistics

Implemented in v2 :
- [X] Return ranked 1/0 in the json_extended
- [X] Link to info/details page of the torrent
- [X] Return pubdate/up date in UTC
- [X] Remove the 10000 limit on search_string when not utilizing imdb/tvdb/tvrage
- [X] Support OR in search with {{something|otherthing}}
- [X] Better , more descriptive error texts
- [X] Response only with json objects by default ( no plain text , no torrents without object )
- [X] Support category strings , use tv instead of 18;41
- [X] Lower the time limit for requests to 1req/2s instead of  1req/10s
- [X] Return pubdate/up date with timezone
- [X] Return error codes with error texts
- [X] Return full property names ex. s->seeders , l->leechers & etc
- [X] Return episode info - imdb/tvrage/tvdb + air date + ep title
- [X] Add category int filtering
- [X] Remake api docs to be more readable + more examples

**Comment by zone117x** [comment]
I'm assuming @Taloth is included for getting RARBG integrated into Sonarr? If so then I'd yield my suggestions to him. If you can get Sonarr integration then there's no point in having Jackette support.

**Comment by rarbg** [comment]
@zone117x 
Yes he is integrating it https://github.com/Sonarr/Sonarr/pull/562
Jackette might support other indexers in the future , thats why you are included :)

**Comment by SchizoDuckie** [comment]
My only complaint so far is that I hit the rate limit during testing. Other
than that i'm pretty happy!

**Comment by Taloth** [comment]
* Include timezone in pubdate, I hate guessing :smile:.
* Use errorcode in conjunction with error text.
* Use full property names 'seeders' instead of 's'. gzip already helps on the compression front.
* Add v1 to the api endpoint, so you can create new versions with breaking changes while preserving the and old api adapter.

For me the 10000 limit doesn't seem a problem, but maybe that's coz we prefer to search by tvdb/imdb/rageids, which imho is much better for the server anyway.

At one point it would be great if the server actually parses and indexes the season/episode/airdate info.

Major problem is always parsing the releases to attach the appropriate tvrage/tvdb id to em.
Already mentioned this to zone a few times, but I see a big difference between a filtered rss feed like newznab/torznab and a full fledged json api.

**Comment by rarbg** [comment]
@SchizoDuckie @Taloth 
Updated the task list
@Taloth 
Dont worry every api version iteration will have different endpoints for backward compatibility

**Comment by SchizoDuckie** [comment]
@rarbg I don't know if you're behind a CDN currently, but I actively abort search requests for find-while-you type with a 500ms delay. So if a equest takes longer than 500ms, and a user has issued an updated search query, a new request will come in and the old one is aborted. You may or may not want to handle those on your serverside for performance. (php's ignore_user_abort equiv)

**Comment by rarbg** [comment]
@SchizoDuckie 
:P Our web servers are handling up to 16000 req/s at peak. I dont see how some small amount of aborted requests can hit our performance

**Comment by SchizoDuckie** [comment]
@rarbg just mentioning for courtesy reasons :)

**Comment by rarbg** [comment]
@SchizoDuckie 
Thank you , but dont worry about it :)

**Comment by Taloth** [comment]
Is the idea to use this api spec purely for rarbg, or also other sites?

Personally I think the torrent community could really benefit from an indexer framework that parses the titles and does all the tvrage/tvdb/numbering stuff.
There are countless trackers that lack that capability and where the admin simply doesn't have the technical know-how or time to make it happen.

**Comment by rarbg** [comment]
the idea of v2 is to make it work for rarbg with all the features you want. if you want a spec that you all agree upon - its not an issue for me to code it in that spec ;)
the "countless trackers" dont have any torrents which are not already available on other sites -  so there is no point of integrating them. also admins that lack the technical knowledge of coding a simple api by default operate sites that are unreliable :)

**Comment by Taloth** [comment]
Was just trying to get a feel for the scope.

My second paragraph wasn't about api, but more about parsing/analysis for indexing purposes, which is a lot harder.

**Comment by rarbg** [comment]
@Taloth 
I can easily parse any release of the top 1000 most popular tv shows to ep info(air date, tvrage & etc ) if thats what you are asking ?

**Comment by rarbg** [comment]
@abeloin , @fernandog , @SchizoDuckie , @Taloth , @JackDandy , @cvium , @zone117x
Im done with almost all the changes , except one for the stats.
Do you have any other requests ?

**Comment by cvium** [comment]
Maybe a field indicating whether it is a ranked/internal release?

**Comment by Taloth** [comment]
I pushed v1 to our develop branch for some feedback and testing yesterday, first comment: Request for a "infoUrl" to the website. :confused:

**Comment by rarbg** [comment]
Added both to the task list

**Comment by rarbg** [comment]
Done!
@Taloth Can you test pubapi_v2.php ?

**Comment by Bon3** [comment]
@rarbg @Taloth Maybe one of you can give me a quick setup guide for this?? I tried reading the docs posted earlier but i can't seem to connect :(

Tried with: https://torrentapi.org/pubapi.php?mode=search&search_imdb=tt123456&category=18;41&token=xxxx

ofc editing the token xxx with the one generated by token generator :( my Sonarr just says failed to connect when i test??

**Comment by Taloth** [comment]
@rarbg 
Tnx for the info_page.
Also, the consistent json object will make deserialization easier for us.

Couple of things:
- ```get_token``` without ```&format=json``` returns raw text but ```Content-Type: application/json```.
- If you want to force json all around, then maybe change ```&format=json_extended```to ```&extended=1``` and get rid of the ```&format=``` ```&response_type=``` params.
- ```pubdate``` still missing timezone. Add ``` +0000``` if it's UTC.

Cosmetics:
- ```filename``` -> ```title``` is better, imho.
- ```seeders/leechers/size``` prolly should be integers instead of string. (same for stuff in episode_info)

Content:
Jimmy.Kimmel.2015.05.13.Elizabeth.Banks.HDTV.x264-CROOKS[rartv]
```episode_info: false``` should be ```null```  (I assume the parser was unable to get any details.)

As mentioned, if at one point you can add query params to search for specific seasons/episodes it would probably save load on your database. Until then I'll use search_string.

**Comment by rarbg** [comment]
@Taloth 
Forced to return gettoken always in json. Updated apidocs_v2.txt
Fixed pubdate to add "+0000"
Cosmetics are also fixed :)
Please test :P

**Comment by Taloth** [comment]
Looks good. I'll wait for the feedback of the others.

Update the apidocs_v2.txt: pubapi.php -> pubapi_v2.php

**Comment by cvium** [comment]
@rarbg I have only done some in-browser tests, but it looks pretty good.

**Comment by miigotu** [comment]
@rarbg Is it possible we could get the id in api to use your download.php for torrent files?  If this is something that you want to prevent, there is a way around it atm in v2:
redirect_to_info.php and find it in the returned page

On one hand, it would speed up our snatching, but on the other there is increased load on your end.

**Comment by rarbg** [comment]
@miigotu 
For legal reasons i cant provide download links to .torrent files.
About the "way around it" - try it. You will see that it wont work , there are limits on the redirect_to_info.
Im happy to discuss that , but its not for github if you catch my drift...
Anyways why do you even need the .torrent file directly ? Cant you utilize magnet links or a combination of the publicly available torrent hosts sites like torrage?

**Comment by miigotu** [comment]
I don't need it, was just making sure someone could not work around it not being there and cause you some trouble down the road pretty much xD.

We already use zoink,torrage,torcache atm. Looping through them just creates a slight more overhead from consecutive requests on failure. Works just fine though.

Right now I'm more working on fixing up SR so that it only has to hit the api once for each series on each refresh, where as now it calls api for each episode. This would drastically reduce processing and bandwidth hogging.

BTW, the proxy thing is fixed now too.

**Comment by rarbg** [comment]
@miigotu
Nice :) Well , thanks
Ill finish up v2 in 2-3 days. We are doing a huge database migration , so as soon as we are ready ill finish it up and everyone can move to v2 :)

**Comment by Taloth** [comment]
FYI, I already updated my code to v2 about a week ago, but didn't push it yet to our develop-branch users, wanted to wait for v2 to be definitive :smile: Not in a hurry either.

**Comment by rarbg** [comment]
@abeloin , @fernandog , @SchizoDuckie , @Taloth , @JackDandy , @cvium , @zone117x , @mancuniancol, @romanvm

v2 is now in production - http://torrentapi.org/apidocs_v2.txt
Please start testing and moving your apps to it :)
What do you think is a good deprecation date for v1 ? 90 days?

**Comment by Hellowlol** [comment]
@rarbg I was wondering if its ok if i use your api in htpc-manager?

**Comment by rarbg** [comment]
@Hellowlol 
Ofcourse you can use it :) Ill add you to the list of projects

**Comment by cvium** [comment]
@rarbg Next release of Flexget will use the new API.

**Comment by miigotu** [comment]
I'll make the switch tonight.

**Comment by rarbg** [comment]
@cvium  , @miigotu 
Please add app_id to your requests , i updated the docs to include it :
When querying the api please use the app_id parameter with your app name so we can generate statistics for it.

**Comment by miigotu** [comment]
@rarbg will do.

**Comment by SchizoDuckie** [comment]
Nearly done with my v2 implementation for DuckieTV. Query results are very odd though: 

![image](https://cloud.githubusercontent.com/assets/111710/8190895/b04e68f6-1465-11e5-8ec6-08fc4077120d.png)
Incorrectly sorted, and there are hits for 'House Of Cards' in the results.

**Comment by rarbg** [comment]
@SchizoDuckie 
Ill check it today , about the search ill need to rework it obviously because it searches in actors too

**Comment by Taloth** [comment]
Can't you use the User-Agent instead of app_id? Sonarr already submits it's own User-Agent on all requests.

**Comment by rarbg** [comment]
@Taloth 
That means parsing all user agents , 48 of them unique to understand what app is what

**Comment by Taloth** [comment]
Very well, added to develop.

As for v1 deprecation, as long as it still works I would keep it alive. Eventually the number of requests should drop low enough, assuming you have statistics on that.

**Comment by miigotu** [comment]
Eh, v2 down right now?

**Comment by rarbg** [comment]
@miigotu 
Fixing @SchizoDuckie's issue by rebuilding the indexes. Shouldnt take long

**Comment by miigotu** [comment]
NP, just making sure our code wasn't broken somehow.

**Comment by SchizoDuckie** [comment]
awesome support :+1:  thanks @rarbg !

**Comment by miigotu** [comment]
Agreed @SchizoDuckie +1. rarbg is definitely my go to public tracker now. After I get our rarbg provider just right and caching results better with less requests I might make it a default/preferred provider in SR as long as I can prevent hammering.

**Comment by rarbg** [comment]
@SchizoDuckie 
Fixed
@SchizoDuckie , @miigotu 
Thanks for the nice words :)
@miigotu 
v2 is up :)

**Comment by SchizoDuckie** [comment]
Works like a charm now. perfect.

**Comment by rarbg** [comment]
Great

**Comment by mancuniancol** [comment]
@rarbg, I've done some test and it is really easy to update my app to the api.  However, I'm forced to use Kodi python version and I get SSL error using urllib2 and requests.  It seems that it is a known issue, but I will try to search more to be able to get the information from the api.  I have the new code ready, but I cannot release it until I could fix the that problem.  Btw, excellent job.

**Comment by rarbg** [comment]
@mancuniancol 
Its a very known issue. Just keep in mind that openssl needs to be latest version and it will work.
However i assume kodi is running mostly on embeded devices so i suggest you make http instead of https requests. The api data is not sensitive at all anyways and there is no way to update openssl on a huge range of embeded devices - like nas & etc. if you really want to use https - fallback to http if https doesnt work on the device :)

**Comment by mancuniancol** [comment]
@rarbg, I didn't try before with http.  Now, it is working :)  thanks.

**Comment by rarbg** [comment]
@cvium 
Dont forget that the api req limits were dropped to 2 secs , so you can modify your set_domain_delay :)

**Comment by SchizoDuckie** [comment]
@rarbg dunno if this makes sense and is possible: could you maybe alter the api req limits to something like 5 searches every 10 seconds? It'd be sweet interaction wise, if you could do some quick find-as-you-type bursts. Doesn't really matter if you have to wait a couple of seconds after that.

**Comment by rarbg** [comment]
@SchizoDuckie 
Right now im keeping just 1 timestamp per token.
Ill have to rework the db to keep token request history and just query db for the last 10 seconds for that token , count and if its > 10 error
Its doable but i'll do it in the next 24-48 hours because i really dont want to modify it right now :)

**Comment by miigotu** [comment]
@SchizoDuckie the limit is to force us to fix our code and not hammer the api. I think after all apps using the api are updated accordingly he might consider it, but for now I would request he doesn't until we know we aren't spamming it.

**Comment by rarbg** [comment]
@miigotu 
About your code ... i was thinking yesterday , would you be interested in search_imdb(tvrage/tvdb)_multi where you can pass multiple imdb/tvrage/tvdb values separated with a character ?

**Comment by SchizoDuckie** [comment]
Yeah, come to think of it, I need to build in a limiter anyway since the
DuckieTV autodownload could produce N results for today. I hereby retract
my request!

**Comment by rarbg** [comment]
@SchizoDuckie
Retraction accepted. However the fine print at the bottom of the README states that anyone retracting requests should pay $5 to every contributor in the project. How do you want to handle payment ? Is bitcoin ok ? :)

**Comment by SchizoDuckie** [comment]
@rarbg sorry, my local currency is pieces of dried out bread. There is no
bitcoin exchange for that yet

**Comment by Hellowlol** [comment]
I likes the suggestion about x api calls in x sec instead of hardcode of 1/2s. Thanks for the api! it's awesome.

**Comment by cvium** [comment]
Having 1 api call per X seconds is easier to handle gracefully though.

**Comment by Hellowlol** [comment]
True.

**Comment by miigotu** [comment]
@rarbg yes, when I get around to reworking the provider it would be nice, but i'm not sure if it is necessary or how to use it with caching cleanly on my end off the top of my head. 
I couldnt just add all of the show ids to one request or (1)the string would be too long, and (2)the results would be hella large.

With my current plan, we will only be hitting the api once for each show max per 15 mins, and caching all results for a show, rather than hitting it once for each episode, so it will be about 10% of the current cost.

Can't I already pass say "search_tvdb=xxxxxx&search_string=s01e01|s03e10|Season%203" ?
Filtering also by a search string would only be a good idea for shows that have like 900 episodes and multiple releases of each lol.

Either way, I can already significantly reduce the spam.

**Comment by rarbg** [comment]
@miigotu 
Eh now i understand your issue. Forget about the search ! Its pointless in your case !
If you are going to index - Why dont you just hit the api without a search , in mode=list and actually do json_extended , from there you will get in the json object episode_info imdb , tvrage , tvdb ids for every upload. Then you can index :) No need for search or anything. Just match if the imdb/tvrage/tvdb is in the user's shows ?

**Comment by miigotu** [comment]
@rarbg it already does, for daily search thread. backlog searches for specific episodes/shows if they are not in the cached list, because the list mode cannot possibly return every possible torrent in your DB.

**Comment by rarbg** [comment]
@miigotu 
For the daily you'll need to to the daily , yep :)
It might be actually cool if your apps supported notifying from my side.
Instead of you constantly asking for "new things" , since most of the htpc boxes are online 24/7 , it would be cool if they can just idle in irc and collect notifications :)

**Comment by miigotu** [comment]
@rarbg Someday in the distant future, I have ideas for ftp providers and irc trolling xD. For now, just focused on working out our list of bugs and situations that are conducive to ebkac.

**Comment by Hellowlol** [comment]
@rarbg Any change you can add more to json_extend while using a using search_string: like codec, resolution etc (i cant remember the rest for the categories from rargb and i cant check from work)?

The "code" (ripped from sr) is done, just waiting on a icon :)

**Comment by miigotu** [comment]
@Hellowlol SR code for rarbg is going to be getting a major overhaul soon. =P

**Comment by Hellowlol** [comment]
No worries ill see if you add anything usefull. I dont need to reenvent the wheel. My needs are pretty basic, just a simple search string and grab as much info as possible. Im just happy i dont have to scape the webpage ^^

**Comment by rarbg** [comment]
@Hellowlol 
I didnt get the category relation with the codec thing ?
Codecs are only 2 on the site - xvid & x264
Resolutions are 3 - SD , 720p, 1080p

Do you want specific codec info - like bitrate & etc ? We keep them for every release including subs streams & etc but it will be very hard to do :)

**Comment by Hellowlol** [comment]
Just forget the thing about categories.

I don't need details like bitrate etc,  but if there was a easy way for you to add source (ex HDTV, WEB-DL), container (mp4, mkv), codec (SD, 7209, 1080p) to extra info returned from json_extended i would appreciate it. 

Don't do it if its alot of work, i can always just use N/A like i have done with the other trackers that don't return that info.

**Comment by rarbg** [comment]
everything is doable
easy - source , codec , resolution
container - hard , need to parse mediainfo for every release :)

**Comment by Hellowlol** [comment]
Ok, drop container then ;)

**Comment by rarbg** [comment]
@Hellowlol 
The cool thing about having a standardized naming convention on your site is that i can just give you a function that you run against the filenames returned from the API and you will get everything you requested :P And yes im lazy

def find_info(filename):
    codec = 'x264';
    v = re.search("(?i)(x265|h265|h\.265)",filename);
    if v:
        codec = 'x265';
    v = re.search("(?i)(xvid)",filename);
    if v:
        codec = 'xvid';

    resolution = 'sd';
    a = re.search("(?i)(720p)",filename);
    if a:
        resolution = '720p';
    a = re.search("(?i)(1080p)",filename);
    if a:
        resolution = '1080p';

    source = 'hdtv';
    s = re.search("(?i)(WEB-DL|WEB_DL|WEB\.DL)",filename);
    if s:
        source = 'web-dl';
    s = re.search("(?i)(WEBRIP)",filename);
    if s:
        source = 'webrip';

    return_info = []
    return_info.append(codec);
    return_info.append(resolution);
    return_info.append(source);
    return return_info;

**Comment by Hellowlol** [comment]
And why didnt i think of that... Thanks!

**Comment by erickok** [comment]
Awesome project guys, really great to see a forward-thinking tracker with a proper API. I added support in my little library erickok/transdroid-search#25 such that RARBG searches are natively supported when using Transdroid or tTorrent on Android.

**Comment by erickok** [comment]
:+1:

**Comment by rarbg** [comment]
Welcome to the project @erickok :)

**Comment by fernandog** [comment]
@rarbg as you are planningg to add code is it possible to add a few more parameters to API Search?

Codec (xvid|h264|h265)
Min size
max size
always in MB

This way we can add more options so user could get only torrents with max size of 1200 mb and min size of 350 for example.

What do you think?

**Comment by rarbg** [comment]
@fernandog 
:D
1. This spec for v2 was open for 20+ days before going to a final stage. Where the fuck were you :D
2. Codec is always xvid or x264. x265 is not mature and its producing shitty results for now. If you want to add it as an option just append it as a search string ? xvid or 264 ?
3. Size limits are stupid , there are no guarantees , especially on HDTV scene releases that the size will be within a range that the user wants and will 99% skip releases :) Why would you want that ? Can you give an example ?

**Comment by fernandog** [comment]
Sorry. 9m baby! Hehehe. Cant code him to sleep.
Sickrage users ask this all the time.
I dont agree too but just wondering

**Comment by rarbg** [comment]
Whats stopping you from filtering them yourself ?
Both size and codec ?

**Comment by fernandog** [comment]
ok! saw the size in json_extended

**Comment by miigotu** [comment]
@fernandog I plan on using json_extended anyways, we need size for failed handling =P. Try no to touch the providers much over the next few days as I have to mod almost every provider to fix proper finder.

**Comment by romanvm** [comment]
Are there any pagination for search results (like "page=2" in a search query string)?  Currently the API returns up to 100 results, but if I search a specific TV show using IMDB ID and the show is running for some time now, I'd like to have access to earlier episodes than included in the first 100.

**Comment by miigotu** [comment]
Just add a search string too, to filter by season
`mode=search&search_imdb=tt123456&category=tv&token=xxxx&search_string=S02E|Season 2`

**Comment by rarbg** [comment]
OR in search is {{something|otherthing}}
so search_string={{S02|Season 2}}

**Comment by miigotu** [comment]
Edited. I haven't even tried that yet, was off the top of my head lol. Just how I planned part of the rewrite.

**Comment by Hellowlol** [comment]
@rarbg rarbg is added :)

**Comment by rarbg** [comment]
@Hellowlol 
Great! Thanks

**Comment by KrunchMuffin** [comment]
Anyone feel like testing Rarbg implementation for CouchPotato?

I have only tested it on Windows. Seems to be OK. First time using Python.

https://github.com/KrunchMuffin/CouchPotatoServer/tree/Develop

**Comment by fernandog** [comment]
@KrunchMuffin can I apply this patch on master?

Or must be applied on develop?

https://github.com/KrunchMuffin/CouchPotatoServer/commit/86bbc87da97a59e782344e345de5457f29fc88e6.patch

**Comment by KrunchMuffin** [comment]
don't see any reason u can't do it on master

**Comment by fernandog** [comment]
@KrunchMuffin testing right now. forced a search in all my wanted movies

got this error. rarbg is the only provider enabled. Lots of this:
```
06-23 19:38:53ERROR[   couchpotato.core.event] Error in event "provider.search.torrent_magnet.movie", that wasn't caught: Traceback (most recent call last):
  File "/home/osmc/CouchPotatoServer/couchpotato/core/event.py", line 15, in runHandler
    return handler(*args, **kwargs)
  File "/home/osmc/CouchPotatoServer/couchpotato/core/media/_base/providers/base.py", line 214, in search
    self._search(media, quality, results)
  File "/home/osmc/CouchPotatoServer/couchpotato/core/media/_base/providers/torrent/rarbg.py", line 96, in _search
    raise ValueError("There is an error in the returned JSON")
ValueError: There is an error in the returned JSON
encoding=UTF-8 debug=True args=['--daemon', '--data_dir', '/home/osmc/.couchpotato/', '--config_file', '/home/osmc/.couchpotato/settings.conf'] app_dir=/home/osmc/CouchPotatoServer data_dir=/home/osmc/.couchpotato desktop=None options=Namespace(config_file='/home/osmc/.couchpotato/settings.conf', console_log=False, daemon=True, data_dir='/home/osmc/.couchpotato/', debug=False, pid_file='/home/osmc/.couchpotato/couchpotato.pid', quiet=False)
```

**Comment by KrunchMuffin** [comment]
Can u paste the search URL? Should be in the log also.

**Comment by fernandog** [comment]
@KrunchMuffin I'm trying removing the https from the gettoken. wait a bit

**Comment by fernandog** [comment]
@KrunchMuffin 

06-23 19:38:12INFO[hpotato.core.plugins.base] Opening url: get https://torrentapi.org/pubapi_v2.php?token=zkf79swm3u&mode=search&search_imdb=tt2395427&min_seeders=1&min_leechers=0&ranked=1&category=movies&format=json_extended, data: []

**Comment by fernandog** [comment]
@rarbg why no results found is classified as an error?

{"error":"No results found","error_code":20}

**Comment by KrunchMuffin** [comment]
ya, would be nice if there was a json attribute with total results count. but no big deal. I will get that fixed tho.

**Comment by rarbg** [comment]
@fernandog 
Thats how the api is from version 1 :D

**Comment by erickok** [comment]
Yes and now that apps rely on it, do we want to change it again? (For my
instance I can still go back and change it for a release, but there might
be instances where thin would break the behaviour.)

**Comment by rarbg** [comment]
@erickok 
Nothing is changed to the api responses errors since version 1.
Nothing is changed to the way it replies to no results since you implemented it.
You didnt read right :)

**Comment by erickok** [comment]
No no, I understand. I was just voting against changing it. :)

**Comment by miigotu** [comment]
@rarbg You have a list of error codes we can check against? Doesnt really make a difference for SR for now, but it might be helpful down the road.

**Comment by rarbg** [comment]
@miigotu 

Cant find imdb in database. Are you sure this imdb exists? 10 
Cant find tvrage/imdb/tvdb in database. Are you sure it exists? 8 
Cant find thetvdb in database. Are you sure this thetvdb exists? 14 
Cant find tvrage in database. Are you sure this tvrage exists? 12 
Invalid character in category 30 
Invalid imdb 9 
Invalid mode 15 
Invalid sort 18 
Invalid token set! 2 
Invalid token set! 3 
Invalid token. Use get_token for a new one! 4 
Invalid tvdb 13 
Invalid tvrage 11 
Invalid value for min_leechers 17 
Invalid value for min_seeders 16 
No results found 20 
No token set! 1 
Too many requests per second. Maximum requests allowed are 1req/2sec Please try again later! 5 
6 , 7 are invalid search pregs on imdb/tvrage/tvdb

Internal if something really fucks up/breaks on our side :

Error occured. Please contact support 21 
Error occured. Please contact support 22 
Error occured. Please contact support 24 
cant execute db queries 19

**Comment by rarbg** [comment]
@erickok
Nothing will be changed without at least a 51% agreement from the people in this ticket :)

**Comment by fernandog** [comment]
@KrunchMuffin when you fix, let me know so I can test. thanks!

**Comment by KrunchMuffin** [comment]
np, I made a bunch of changes for better performance and fixed the issue with only 1 result appearing, if u noticed that. Hopefully get it pushed tonight.

**Comment by Hellowlol** [comment]
@rarbg can you add thouse responses to the api doc. Its nice to have everything in the same place

**Comment by erickok** [comment]
Might also be good to add in the doc that it is mandatory to set a User-Agent. Otherwise I received HTTP errors.

**Comment by KrunchMuffin** [comment]
@fernandog latest code commited

https://github.com/KrunchMuffin/CouchPotatoServer/tree/Develop

let me know. thanks.

**Comment by fernandog** [comment]
@KrunchMuffin seems ok. No errors

[e.providers.torrent.rarbg] No results returned from Rarbg

Didnt test the snatch as nothing available in mu wanted list

**Comment by fernandog** [comment]
@KrunchMuffin notifications didn't worked in CP UI (the little "eye") when snatched. 
can you confirm?

also can you add the IMDB tt in this message?
06-26 09:43:06ERROR[e.providers.torrent.rarbg] There is an error in the returned JSON: Cant find imdb in database. Are you sure this imdb exists?

Movie was added correctly to CP so it's strange that it can find  the imdb

**Comment by KrunchMuffin** [comment]
ya, it's working for me, but I don't even touch that code. That is part of the main program.

I can see if I can return the imdb code, not sure right now. Be best if the rarbg api returned it with that error. what was the name you were searching?

**Comment by fernandog** [comment]
full search for all wanted. 40 items. 10 returned with this error

**Comment by KrunchMuffin** [comment]
can u give me some names to test?

**Comment by fernandog** [comment]
@rarbg @KrunchMuffin 

https://torrentapi.org/pubapi_v2.php?token=o92sypktwg&mode=search&search_imdb=tt2398241&min_seeders=1&min_leechers=0&ranked=1&category=movies&format=json_extended

{"error":"Cant find imdb in database. Are you sure this imdb exists?","error_code":10}

http://www.imdb.com/title/tt2398241/

**Comment by KrunchMuffin** [comment]
ya, looks right. The main search is done by CP, not my code. CP just hands me the info and I pass it on to rarbg. These are the links CP checks and they all return no results for that movie. Assuming cause it's 2 years in the future, but I don't know. Does it actually show up in your wanted list? if so, very weird and may want to log a CP bug.

06-26 09:56:21 INFO [0m[hpotato.core.plugins.base] Opening url: get https://api.couchpota.to/search/SEE+RANK+Get+Smurfy/?limit=5, data: [][0m
06-26 09:56:21 INFO [0m[hpotato.core.plugins.base] Opening url: get http://www.omdbapi.com/?type=movie&y=&t=SEE+RANK+Get+Smurfy, data: [][0m
06-26 09:56:21 INFO [0m[hpotato.core.plugins.base] Opening url: get http://api.themoviedb.org/3/search/movie?api_key=9b939aee0aaafc12a65bf448e4af9543&search_type=ngram&query=SEE+RANK+Get+Smurfy,

**Comment by fernandog** [comment]
yes. I add my movies to trakt. and CP syncs with trakt and add to the wanted list. There's tons of future movies.

**Comment by miigotu** [comment]
@KrunchMuffin you could consider that a bug in either or both rarbg api and CP handling of error codes.

Either rarbg should just return an empty result when the imdb|tvrage|tvdb_id is not found, or CP should a.) not be searching for movies that release in the furture or b.) handle the error code better.

**Comment by KrunchMuffin** [comment]
Well CP already does do that. If u do a search at the top for a movie in the future, it does what it should and returns no results so it has no I'd to send to rarbg etc. so it's prolly the trakt integration.

**Comment by miigotu** [comment]
@fernandog the request should never be made to rarbg if the movie date is in the future. 100% needs to be fixed somewhere in cp. Actually, the error should be handled, and it should not be searched.

**Comment by fernandog** [comment]
@miigotu CP has this option (mine is enabled)

**Always Search**
Search for movies even before there is a ETA. Enabling this will probably get you a lot of fakes.

**Comment by fernandog** [comment]
@rarbg did you see this? https://github.com/rarbg/torrentapi/issues/1#issuecomment-115699871

How often and what's the "scope" your imdb database updates?

**Comment by rarbg** [comment]
handle the error. our imdb database usually doesnt include titltes which were never released on the site

**Comment by fernandog** [comment]
Thanks
@KrunchMuffin so all indexers errors should be 'no results found' agree?

**Comment by KrunchMuffin** [comment]
all i get back from CP is a imdbid. I don't get specifics. if there is no id, no search in rarbg is done.

**Comment by miigotu** [comment]
@KrunchMuffin Just add the check here: https://github.com/KrunchMuffin/CouchPotatoServer/blob/Develop/couchpotato/core/media/_base/providers/torrent/rarbg.py#L112

```
if data['error_code'] in range(5, 15):
    log.debug("Not found or Invalid ID")
```
Or you could break them into separate error messages based on the error code, and only log errors for actual errors.

**Comment by KrunchMuffin** [comment]
but that is after a rarbg search. I believe everyone is talking about not searching at all?

**Comment by rarbg** [comment]
@KrunchMuffin 
They are saying that other apps have the capability of not searching for future year releases.
If you do have the year of the current imdb you are searching for and the current year is < than this imdb year you shouldnt search for it , simply because searching for a 2017 release in 2016 is waste of bandwith :)

**Comment by KrunchMuffin** [comment]
ahh..ok, so u want me to check year against current year and if >current year, not search. that makes sense and should be able to do that.

**Comment by rarbg** [comment]
@KrunchMuffin 
Also you should add what @miigotu gave you in case the release is not available so the error is handled

**Comment by KrunchMuffin** [comment]
I am doing that, just not looking for specific id. but I can add it.

                if 'error_code' in data:
                    if data['error'] == 'No results found':
                        log.debug("No results returned from Rarbg")
                    else:
                        log.error("There is an error in the returned JSON: %s", data['error'])

**Comment by rarbg** [comment]
@KrunchMuffin 
That error catching wont catch expired api tokens ( 4 )

**Comment by KrunchMuffin** [comment]
I take care of that elsewhere

**Comment by Hellowlol** [comment]
I wouldn't make anything global with generic name like token.

**Comment by KrunchMuffin** [comment]
pushed new code
https://github.com/KrunchMuffin/CouchPotatoServer/tree/Develop

**Comment by rarbg** [comment]
@KrunchMuffin 
You need to send app_id for internal purposes + a valid user agent(not idea what couchpotato's default is ) so cloudflare doesnt trip your requests

**Comment by miigotu** [comment]
@rarbg should it be a standard browser user agent, or like the one we use in SR?

**Comment by rarbg** [comment]
@miigotu 
It shouldnt be a browser agent or curl/python/wget/whatever crawler you can think of/script language default user agent else cf starts acting weird

**Comment by fernandog** [comment]
@KrunchMuffin no errors here.

**Comment by KrunchMuffin** [comment]
Shit, I missed the app_id, sorry about that.
as for the user agent, CP currently sets it to
`user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:34.0) Gecko/20100101 Firefox/34.0'`

I haven't had any issues with CF. But I can make it CouchPotato/1.0 I guess if you think it helps.

**Comment by KrunchMuffin** [comment]
@fernandog I tried to do the squash, dunno if it is correct. I am not a GIT master by any means.

**Comment by fernandog** [comment]
@rarbg Im curious. How many SR users using RARBG? Do you have this stat? Thanks

**Comment by rarbg** [comment]
@fernandog 
we keep stats only for the past 24 hours , for now until i make the app_id stats :
2513 unique ips hit the api with the user agent SickRage*
375915 requests to the api with the user agent SickRage*

**Comment by fernandog** [comment]
RARBG PR to couchtpotato got merged! 
can you all test it too?

**Comment by rarbg** [comment]
@KrunchMuffin 
Thank you for all the hard work!
@fernandog 
Thank you for helping out!

**Comment by KrunchMuffin** [comment]
No problem.

**Comment by fernandog** [comment]
@rarbg can you tell how many users are using RARBG for CP?

**Comment by rarbg** [comment]
@fernandog 
2 ! :)

**Comment by KrunchMuffin** [comment]
I don't think he has merged it into master or the executable so may be a
while before a high adoption rate.

**Comment by fernandog** [comment]
damn! forgot about that =) it's still in develop. Sorry!

**Comment by miigotu** [comment]
Lol @ 2.... Haters.

**Comment by rarbg** [comment]
@KrunchMuffin 
You emailed saying you wanted top 10 rss included in the api :)
Just sort by seeders , ranked 1 ;)

**Comment by rarbg** [comment]
Ok guys.
v1 is going to be deprecated in 90 days starting from today.
v2 is public and used by all active apps.
Im opening a new issue for v3 features & etc
https://github.com/rarbg/torrentapi/issues/3
Closing this one cause it got too huge :)

**Comment by KrunchMuffin** [comment]
No no. Cp has options for charts and stuff and wanted to include urs.
Nothing to do with api that's why I didn't post here.

**Comment by fernandog** [comment]
anyone having issues with torrage giving bad .torrents files?

```
Torrent file content is empty
Content-Encoding was gzip, but content was not compressed. URL: http://torrage.com/torrent/1F5DFD0CD7366639FDB196D60BDD2EF8CC2A219D.torrent
Downloading ************** from Rarbg
Picked ********* as the best
Requested url http://torrage.com/torrent/1F5DFD0CD7366639FDB196D60BDD2EF8CC2A219D.torrent returned status code is 200
```
